### PR TITLE
(Docs) getStaticProps & getServerSideProps TS section (#40607)

### LINKED ISSUE #40607: Docs: getStaticProps & InferGetStaticPropsType | TypeScript
### What is the improvement or update you wish to see?

image

In the documentation, it is advised to type getStaticProps first, and then in the example with InferGetStaticPropsType, getStaticProps is not typed. The documentation should specify this type behavior explicitly. All examples of type behavior are listed in the examples below.

The same with the other rendering methods

### Is there any context that might help us understand?

Example 1. InferGetStaticPropsType is working as expected. Object `posts` in page Posts is type `types.Post[]`.
```
const Posts = ({ posts }: InferGetStaticPropsType<typeof getStaticProps>) => {
    ...
};

export const getStaticProps: GetStaticProps<{ posts: types.Post[] }> = async () => {
    let data: {
        allPosts: types.Post[];
    };

   // fetching data

    const posts = data.allPosts;
    return {
        props: {
            posts,
        },
    };
};
```

Example 2. The same as in the previous example, but the getStaticProps type is specified implicitly. This example is listed in the documentation.
```
const Posts = ({ posts }: InferGetStaticPropsType<typeof getStaticProps>) => {
    ...
};

export const getStaticProps = async () => {
    let data: {
        allPosts: types.Post[];
    };

   // fetching data

    const posts = data.allPosts;
    return {
        props: {
            posts,
        },
    };
};
```

Example 3. InferGetStaticPropsType is not working. Object `posts` in page Posts is type `any`.
```
const Posts = ({ posts }: InferGetStaticPropsType<typeof getStaticProps>) => {
    ...
};

export const getStaticProps: GetStaticProps = async () => {
    let data: {
        allPosts: types.Post[];
    };

   // fetching data

    const posts = data.allPosts;
    return {
        props: {
            posts,
        },
    };
};
```

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/api-reference/data-fetching/get-static-props#getstaticprops-with-typescript

### PR DESCRIPTION
**Pull request by sicikh**
Added examples to `getStaticProps` & `getServerSideProps` explicit and implicit typings for TypeScript.

Related issues: fixes https://github.com/vercel/next.js/issues/40607

### CODE REVIEW

**Comment by HaNdTriX** [comment]
Hey @sicikh  thank you working on improving the documentation. 

From my point of view I don't recommend mixing the type `GetServerSideProps` and `InferGetServerSidePropsType`. 
This is because the type `GetServerSideProps` explicitly tells typescript what `getStaticProps` will return. `InferGetServerSidePropsType` will infer these types instead of defining them. If you now use both  `InferGetServerSidePropsType` will get the type from `GetServerSideProps<T>` instead of inferring it from js.

Screenshot 2022-09-18 at 20 49 01

Screenshot 2022-09-18 at 20 49 29

The same applies to `GetStaticProps`. I am currently working on fixing `InferGetServerSidePropsType` types in https://github.com/vercel/next.js/pull/40635.

**Comment by sicikh** [comment]
> Hey @sicikh thank you for working on improving the documentation 🙏.
> 
> From my point of view I don't recommend mixing the type `GetServerSideProps` and `InferGetServerSidePropsType`. This is because the type `GetServerSideProps` explicitly tells typescript what `getStaticProps` will return. `InferGetServerSidePropsType` will infer these types instead of defining them. If you now use both `InferGetServerSidePropsType` will get the type from `GetServerSideProps<T>` instead of inferring it from js.
> 
> Screenshot 2022-09-18 at 20 49 01
> 
> Screenshot 2022-09-18 at 20 49 29
> 
> As you can see the last approach infers the correct types the first one uses a default type which is critical from my point of view.
> 
> The same applies to `GetStaticProps`. I am currently working on fixing `InferGetServerSidePropsType` types in #40635.

Thanks for the reply! The problem given in your example is exactly what was missing in the existing documentation. The current behavior of types when we explicitly specify the type of functions is the standard behavior of TypeScript with type calculation. If we want to type the props in the page, while maintaining the explicit typing of `getStaticProps`, then we must give the types of the returned props in the generics of the function `getStaticProps: GetStaticProps<{ propsObject: propsObjectType }>`. Otherwise, `InferGetStaticProps` will type the props as `[key:string]: any` because of the following:

image

That is why it is necessary to specify two examples in the documentation, with an explicit and implicit specification of the type of function of data fetching.

File: docs/api-reference/data-fetching/get-server-side-props.md [hunk 1]
```diff
@@ -104,18 +104,53 @@ export async function getServerSideProps(context) {
 
 ### getServerSideProps with TypeScript
 
-For TypeScript, you can use the `GetServerSideProps` type from `next`:
+To specify type of `getServerSideProps` explicitly, you can use `GetServerSideProps` from `next`:
```
Comment (balazsorban44) [inline docs/api-reference/data-fetching/get-server-side-props.md:107]:
```suggestion
The type of `getServerSideProps` be specified using `GetServerSideProps` from `next`:
```

File: docs/api-reference/data-fetching/get-static-props.md [hunk 2]
```diff
@@ -202,18 +202,61 @@ export default Blog
 
 ## getStaticProps with TypeScript
 
-You can use the `GetStaticProps` type from `next` to type the function:
+To specify type of `getStaticProps` explicitly, you can use `GetStaticProps` from `next`:
```
Comment (balazsorban44) [inline docs/api-reference/data-fetching/get-static-props.md:205]:
```suggestion
The type of `getStaticProps` can be specified using `GetStaticProps` from `next`:
```

**Review by balazsorban44 — APPROVED** [review]
Thanks!

File: docs/api-reference/data-fetching/get-static-props.md [hunk 3]
```diff
@@ -202,18 +202,61 @@ export default Blog
...
+    props: {
+      posts,
+    },
+  }
 }
 ```
 
 If you want to get inferred typings for your props, you can use `InferGetStaticPropsType<typeof getStaticProps>`:
 
+```tsx
```
Comment (leerob) [inline docs/api-reference/data-fetching/get-static-props.md:231]: Will tsx work? I am unsure (worth checking)

**Pull request merged**